rust/crd/src/authentication.rs

```diff
@@ -0,0 +1,118 @@
+    /// - If clients need to authenticate themselves against the broker via TLS
+    /// - Which ca.crt to use when validating the provided client certs
+    /// This will override the server TLS settings (if set) in `spec.clusterConfig.tls.serverSecretClass`.
+    pub authentication_class: String,
+}
+
+#[derive(Clone, Debug)]
+/// Helper struct that contains resolved AuthenticationClasses to reduce network API calls.
+pub struct ResolvedAuthenticationClasses {
+    resolved_authentication_classes: Vec<AuthenticationClass>,
+}
+
+impl ResolvedAuthenticationClasses {
+    pub fn new(resolved_authentication_classes: Vec<AuthenticationClass>) -> Self {
+        Self {
+            resolved_authentication_classes,
+        }
+    }
+
+    /// Return the (first) TLS `AuthenticationClass` if available
+    pub fn get_tls_authentication_class(&self) -> Option<&AuthenticationClass> {
+        self.resolved_authentication_classes
+            .iter()
+            .find(|auth| matches!(auth.spec.provider, AuthenticationClassProvider::Tls(_)))
+    }
+
+    /// Validates the resolved AuthenticationClasses.
+    /// Currently errors out if:
+    /// - More than one AuthenticationClass was provided
+    /// - AuthenticationClass mechanism was not supported
+    pub fn validate(&self) -> Result<Self, Error> {
+        if self.resolved_authentication_classes.len() > 1 {
+            return Err(Error::MultipleAuthenticationClassesProvided);
+        }
+
+        for auth_class in &self.resolved_authentication_classes {
+            match &auth_class.spec.provider {
+                AuthenticationClassProvider::Tls(_) => {}
+                _ => {
+                    return Err(Error::AuthenticationMethodNotSupported {
+                        authentication_class: ObjectRef::from_obj(auth_class),
+                        method: auth_class.spec.provider.to_string(),
+                    })
+                }
+            }
+        }
+
+        Ok(self.clone())
+    }
+}
+
+/// Resolve provided AuthenticationClasses via API calls and validate the contents.
+/// Currently errors out if:
+/// - AuthenticationClass could not be resolved
+/// - Validation failed
+pub async fn resolve_authentication_classes(
+    client: &Client,
+    auth_classes: &Vec<KafkaAuthentication>,
+) -> Result<ResolvedAuthenticationClasses, Error> {
+    let mut resolved_authentication_classes: Vec<AuthenticationClass> = vec![];
+
+    for auth_class in auth_classes {
+        resolved_authentication_classes.push(
+            AuthenticationClass::resolve(client, &auth_class.authentication_class)
+                .await
+                .context(AuthenticationClassRetrievalSnafu {
+                    authentication_class: ObjectRef::<AuthenticationClass>::new(
+                        &auth_class.authentication_class,
+                    ),
+                })?,
+        );
+    }
+
+    ResolvedAuthenticationClasses {
+        resolved_authentication_classes,
+    }
```

Comment:
Why not use `ResolvedAuthenticationClasses::new` (or `ResolvedAuthenticationClasses::from_references`) ? ;)

Reply:
done